pkg/statsd/handler_cloud.go

```diff
@@ -107,6 +127,45 @@ func (ch *CloudHandler) WaitForEvents() {
 	ch.events.WaitForEvents()
 }
 
+func (ch *CloudHandler) RunMetrics(ctx context.Context, statser stats.Statser) {
+	// This will be cleaned up later
+	if me, ok := ch.cloud.(MetricEmitter); ok {
+		var wg wait.Group
+		defer wg.Wait()
+		wg.Start(func() {
+			me.RunMetrics(ctx, statser)
+		})
+	}
+
+	// All the channels are unbuffered, so no CSWs
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			ch.emit(statser)
+		}
+	}
+}
+
+func (ch *CloudHandler) emit(statser stats.Statser) {
```

Comment:
As discussed, all these atomic operations worry me a bit. If not hot paths now they could become so in the future with a bit of refactoring. We might be able to replace them with channels between producer and consumer (so that not every operation by the producer is atomic), or else add perf tests so we can see how this changes behaviour vs the current code and regression due to future changes. I'll have a deeper look at it...

Reply:
It is always possible to test if it matters - comment them all out, rebuild and redeploy :) I would be very surprised if they makes any difference.

Reply:
3 of these are with it, 3 are without.  I forget which is which.
```
BenchmarkCloudHandlerDispatchMetric-4            2000000               628 ns/op             704 B/op         16 allocs/op
BenchmarkCloudHandlerDispatchMetric-4            2000000               568 ns/op             704 B/op         16 allocs/op
BenchmarkCloudHandlerDispatchMetric-4            2000000               554 ns/op             704 B/op         16 allocs/op
BenchmarkCloudHandlerDispatchMetric-4            3000000               544 ns/op             704 B/op         16 allocs/op
BenchmarkCloudHandlerDispatchMetric-4            3000000               525 ns/op             704 B/op         16 allocs/op
BenchmarkCloudHandlerDispatchMetric-4            2000000               688 ns/op             704 B/op         16 allocs/op
```

This is on my laptop, it may not be so clear when there's a billion cores hammering it, but I expect it to be fine because normal access takes a read lock on CloudHandler.rw which at its heart is a couple of atomic.AddInt32's.  We can probably reclaim those nanos by fixing line 421.

Reply:
Yes it's always possible to test but in practice if no one runs the tests (because they don't exist) we won't spot regressions. Thanks for adding the benchmarks @tiedotguy